gnome-image-installer/pages/confirm/gis-confirm-page.c

```diff
@@ -0,0 +1,281 @@
+
+  g_signal_connect (priv->cancel_button,
+                    "clicked",
+                    G_CALLBACK (gis_confirm_page_cancel_clicked),
+                    self);
+  g_signal_connect (priv->continue_button,
+                    "clicked",
+                    G_CALLBACK (gis_confirm_page_continue_clicked),
+                    self);
+
+  gtk_widget_show (GTK_WIDGET (self));
+}
+
+static GtkBuilder *
+gis_confirm_page_get_builder (GisPage *page)
+{
+  /* This class uses the newer gtk_widget_class_set_template_from_resource()
+   * mechanism.
+   */
+  return NULL;
+}
+
+static void
+gis_confirm_page_locale_changed (GisPage *page)
+{
+  gis_page_set_title (page, _("Endless OS Unattended Installation"));
+}
+
+static void
+gis_confirm_page_class_init (GisConfirmPageClass *klass)
+{
+  GisPageClass *page_class = GIS_PAGE_CLASS (klass);
+  GObjectClass *object_class = G_OBJECT_CLASS (klass);
+
+  gtk_widget_class_set_template_from_resource (GTK_WIDGET_CLASS (klass), "/org/gnome/initial-setup/gis-confirm-page.ui");
+
+  gtk_widget_class_bind_template_child_private (GTK_WIDGET_CLASS (klass), GisConfirmPage, vendor_label);
+  gtk_widget_class_bind_template_child_private (GTK_WIDGET_CLASS (klass), GisConfirmPage, product_label);
+
+  gtk_widget_class_bind_template_child_private (GTK_WIDGET_CLASS (klass), GisConfirmPage, image1_label);
+  gtk_widget_class_bind_template_child_private (GTK_WIDGET_CLASS (klass), GisConfirmPage, model1_label);
+  gtk_widget_class_bind_template_child_private (GTK_WIDGET_CLASS (klass), GisConfirmPage, device1_label);
+  gtk_widget_class_bind_template_child_private (GTK_WIDGET_CLASS (klass), GisConfirmPage, size1_label);
+
+  gtk_widget_class_bind_template_child_private (GTK_WIDGET_CLASS (klass), GisConfirmPage, warning_box);
+  gtk_widget_class_bind_template_child_private (GTK_WIDGET_CLASS (klass), GisConfirmPage, warning_label);
+
+  gtk_widget_class_bind_template_child_private (GTK_WIDGET_CLASS (klass), GisConfirmPage, wrong_computer_box);
+
+  gtk_widget_class_bind_template_child_private (GTK_WIDGET_CLASS (klass), GisConfirmPage, countdown_label);
+  gtk_widget_class_bind_template_child_private (GTK_WIDGET_CLASS (klass), GisConfirmPage, cancel_button);
+  gtk_widget_class_bind_template_child_private (GTK_WIDGET_CLASS (klass), GisConfirmPage, continue_button);
+
+  page_class->hide_forward_button = TRUE;
+  page_class->hide_backward_button = TRUE;
+  page_class->page_id = PAGE_ID;
+  page_class->get_builder = gis_confirm_page_get_builder;
+  page_class->locale_changed = gis_confirm_page_locale_changed;
+  page_class->shown = gis_confirm_page_shown;
+  object_class->constructed = gis_confirm_page_constructed;
+}
+
+static void
+gis_confirm_page_init (GisConfirmPage *self)
+{
+  g_resources_register (confirm_get_resource ());
+
+  gtk_widget_init_template (GTK_WIDGET (self));
+}
+
+void
+gis_prepare_confirm_page (GisDriver *driver)
+{
+  if (gis_store_is_unattended ())
+    gis_driver_add_page (driver,
```

Comment:
I assume `gis_driver_add_page()` takes ownership of the object you pass it? (I haven’t checked.)

Reply:
It does, sort of (by which I mean, it does not take a new reference, but nor does it release the reference).

Reply:
But it does later unref what it’s passed, right?

Reply:
Well… I think the answer is "yes". `GisPage` is a `GInitiallyUnowned`. It gets added to a `GtkStack`, and in response to `::destroyed` the GisAssistant (which GisDriver just hands the page over to) removes the pointer from the list. So I think the `GtkStack` consumes the initial floating reference and there are no other references.

These are all singletons so I'm not that worried anyway!

Reply:
If you’re satisfied, I’ll leave it at that!